gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessageOperationTransformers.java

```diff
@@ -0,0 +1,122 @@
+import com.google.api.gax.longrunning.OperationSnapshot;
+import com.google.api.gax.rpc.ApiExceptionFactory;
+import com.google.api.gax.rpc.StatusCode.Code;
+
+/**
+ * Transformers from OperationSnapshot wrappers to the underlying native ApiMessage objects. Public
+ * for technical reasons; intended for use by generated code.
+ */
+@BetaApi("The surface for use by generated code is not stable yet and may change in the future.")
+public class ApiMessageOperationTransformers {
+  private ApiMessageOperationTransformers() {}
+
+  public static class ResponseTransformer<ResponseT extends ApiMessage>
+      implements ApiFunction<OperationSnapshot, ResponseT> {
+    private final Class<ResponseT> responseTClass;
+
+    private ResponseTransformer(Class<ResponseT> responseTClass) {
+      this.responseTClass = responseTClass;
+    }
+
+    /** Unwraps an OperationSnapshot and returns the contained method response message. */
+    @SuppressWarnings("unchecked")
+    public ResponseT apply(OperationSnapshot operationSnapshot) {
+      if (!operationSnapshot.getErrorCode().getCode().equals(Code.OK)) {
+        throw ApiExceptionFactory.createException(
+            "Operation with name \""
+                + operationSnapshot.getName()
+                + "\" failed with status = "
+                + operationSnapshot.getErrorCode()
+                + " and message = "
+                + operationSnapshot.getErrorMessage(),
+            null,
+            operationSnapshot.getErrorCode(),
+            false);
+      }
+      if (!operationSnapshot.getResponse().getClass().isAssignableFrom(responseTClass)) {
+        throw ApiExceptionFactory.createException(
+            new Throwable(
+                String.format(
+                    "Operation with name \"%s\" succeeded, but its response type %s cannot be cast to %s.",
+                    operationSnapshot.getName(),
+                    operationSnapshot.getResponse().getClass().getCanonicalName(),
+                    responseTClass.getCanonicalName())),
+            operationSnapshot.getErrorCode(),
+            false);
+      }
+      return (ResponseT) operationSnapshot.getResponse();
+    }
+
+    public static <ResponseT extends ApiMessage>
+        ApiMessageOperationTransformers.ResponseTransformer<ResponseT> create(
+            Class<ResponseT> packedClass) {
+      return new ApiMessageOperationTransformers.ResponseTransformer<>(packedClass);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public static class MetadataTransformer<MetadataT extends ApiMessage>
+      implements ApiFunction<OperationSnapshot, MetadataT> {
+    private final Class<MetadataT> metadataTClass;
+
+    private MetadataTransformer(Class<MetadataT> metadataTClass) {
+      this.metadataTClass = metadataTClass;
+    }
+
+    /** Unwraps an OperationSnapshot and returns the contained operation metadata message. */
+    @Override
+    public MetadataT apply(OperationSnapshot operationSnapshot) {
```

Comment:
It seems like this is same logic as in the response transformer (the "response type" and "metadata type" in error message string can be a parameter). Please consider putting this in helper method and reuse it (static or make transformers extend same class and put this method in the parent class).

Reply:
Done; implemented in `transformEntityFromOperationSnapshot()`